File: AAS.Tools/Types/Results/DataResult.cs
namespace AAS.Tools.Types.Results;

public readonly struct DataResult<T>
{
    public T? Data { get; }
    public Error[] Errors { get; }
    public bool IsSuccess => !Errors?.Any() ?? true;

    public DataResult(T data, IEnumerable<Error>? errors = null)
    {
        Data = data;
        Errors = errors?.ToArray() ?? new Error[0];
    }

    public static DataResult<T> Success(T data)
    {
        return new DataResult<T>(data);
    }

    public static DataResult<T?> Fail(IEnumerable<Error> errors)
    {
        return new DataResult<T?>(default, errors);
    }

    public static DataResult<T?> Fail(string error)
    {
        return new DataResult<T?>(default, new[] { new Error(error) });
    }
}

File: AAS.BackOffice/client_app/src/tools/results/dataResult.ts
import { ErrorData, mapToErrorData } from './errorData';

export class DataResult<T> {
    public isSuccess = this.errors.length === 0;

    constructor(
        public data: T | null,
        public errors: ErrorData[]
    ) { }

    public static success<T>(value: T): DataResult<T> {
        return new DataResult<T>(value, []);
    }

    public static fail<T>(errors: ErrorData[]): DataResult<T> {
        return new DataResult<T>(null, errors);
    }

    public getErrorsString = (): string => {
        return this.errors.map(error => error.message).join('. ')
    }
}

export function mapToDataResult<T>(result: any): DataResult<T> {
    if (!result.isSuccess) return DataResult.fail<T>(mapToErrorData(result.errors))

    return DataResult.success(result.data);
}
